unified-runtime/source/adapters/level_zero/v2/queue_immediate_in_order.cpp

```diff
@@ -705,31 +707,153 @@ ur_result_t ur_queue_immediate_in_order_t::enqueueWriteHostPipe(
   return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
 }
 
+ur_result_t ur_queue_immediate_in_order_t::enqueueUSMAllocHelper(
+    ur_usm_pool_handle_t pPool, const size_t size,
+    const ur_exp_async_usm_alloc_properties_t *, uint32_t numEventsInWaitList,
+    const ur_event_handle_t *phEventWaitList, void **ppMem,
+    ur_event_handle_t *phEvent, ur_usm_type_t type) {
+  auto commandListLocked = commandListManager.lock();
+
+  if (!pPool) {
+    pPool = hContext->getAsyncPool();
+  }
+
+  auto device = (type == UR_USM_TYPE_HOST) ? nullptr : hDevice;
+
+  ur_event_handle_t originAllocEvent = nullptr;
+  auto asyncAlloc = pPool->allocateEnqueued(hContext, this, true, device,
+                                            nullptr, type, size);
+  if (!asyncAlloc) {
+    auto Ret = pPool->allocate(hContext, device, nullptr, type, size, ppMem);
+    if (Ret) {
+      return Ret;
+    }
+  } else {
+    std::tie(*ppMem, originAllocEvent) = *asyncAlloc;
+  }
+
+  auto waitListView = getWaitListView(commandListLocked, phEventWaitList,
+                                      numEventsInWaitList, originAllocEvent);
+
+  ur_command_t commandType = UR_COMMAND_FORCE_UINT32;
+  switch (type) {
+  case UR_USM_TYPE_HOST:
+    commandType = UR_COMMAND_ENQUEUE_USM_HOST_ALLOC_EXP;
+    break;
+  case UR_USM_TYPE_DEVICE:
+    commandType = UR_COMMAND_ENQUEUE_USM_DEVICE_ALLOC_EXP;
+    break;
+  case UR_USM_TYPE_SHARED:
+    commandType = UR_COMMAND_ENQUEUE_USM_SHARED_ALLOC_EXP;
+    break;
+  default:
+    UR_LOG(ERR, "enqueueUSMAllocHelper: unsupported USM type");
+    throw UR_RESULT_ERROR_UNKNOWN;
+  }
+
+  auto zeSignalEvent = getSignalEvent(commandListLocked, phEvent, commandType);
+  auto [pWaitEvents, numWaitEvents] = waitListView;
+
+  if (phEvent) {
+    (*phEvent)->setOriginAllocEvent(originAllocEvent);
```

Comment:
Do we do anything with event->originAllocEvent after setting it? Or is it just to prolong it's lifetime?

Reply:
Yes, it's there to prolong the lifetime.

Reply:
I'm not sure this is really needed then - in v2 we use counter-based events which can be safely released and reused immediately after they are passed to an enqueue command. What would happen if we just released that event right here?

Reply:
I thought that we cannot reuse the event immediately after wait command is appended to the command list. As that's not the case, then yes, it's not needed.